platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -17,68 +19,366 @@ components within the platform.
+config.txt a poor medium for carrying out `Roster` updates. Writing the updated roster out to disk is less secure
+than storing it in the state. When TSS is implemented, there will be TSS related key material stored in the state,
+and it is imperative that this key material stay in sync with the `Roster` data. Leaving the content as a modifiable
+file on disk would pose risks of introducing errors.
 
-### Roster Public API
+## Changes
 
-#### Setting the Candidate Roster
+The platform `AddressBook` lifecycle is being replaced with a new `roster` lifecycle. A new feature flag
+`addressbook.useRosterLifecycle` will be used to toggle between the current `config.txt` based address book
+lifecycle and the new `roster` lifecycle. The feature flag will be set to `false` by default. The new roster
+lifecycle will be enabled by setting the flag to `true`.
 
-A new method will be added to the platform state to allow the App to submit a Candidate Roster to the platform:
+### Architecture and/or Components
 
-```java
-//in PlaformState
+No new components are proposed. Without loss of capability, the platform `AddressBook` data structure is replaced in
+all platform components with the `Roster` data structure, including in the `ConsensusRound` data structure that is
+passed to the application. If the application is dependent on any of the removed fields on an address in the
+`AddressBook` data structure, it will have to look that information up from its own data storage for the Hedera Address
+Book.
 
-void setCandidateRoster(@NonNull final Roster candidateRoster);
-```
+Updating the roster is now performed through writing a `candidate` roster in the state instead of writing a new
+`config.txt` and the associated gossip certificates to disk.
+
+#### Roster Types.
+
+This proposal introduces four types of `Roster` - `active`, `candidate`, `genesis`, and `override`.
+
+1. `Active Rosters` are paired with the round numbers for when they became active and are used to validate event
+   signatures. The active roster with latest round number is used to determine consensus
+   * The `Roster History` is a list of active rosters paired with the rounds they became active. At this time the
+     history length is at most 2.
+     * `currentRoster` - the active roster in the history paired with the highest round number.
+     * `previousRoster` - the active roster in the history paired with the lowest round number.
+     * In this document we denote the roster history as an ordered list of pairs by descending round
+       number. Example:  [(currentRoster, 2), (previousRoster, 1)].
+2. A `Genesis Roster` is loaded from disk at the genesis of a new network and becomes the first active roster.
+3. The `Candidate Roster` is a roster that will become active at the next software upgrade.
+4. An `Override Roster` is loaded from disk and unconditionally set as the latest active roster. This is used to
```

Comment:
Could the override roster also be used in production in case of an emergency, for example if the roster was corrupted and the process to update it had a bug that prevented it from being fixed with transactions?

Reply:
yes, but this will break TSS once we have TSS.